test/core/pin.js

```diff
@@ -0,0 +1,328 @@
+              { type: 'recursive',
+                hash: 'QmTAMavb995EHErSrKo7mB8dYkpaSJxu6ys1a6XJyB2sys' },
+              { type: 'indirect',
+                hash: 'QmTMbkDfvHwq3Aup6Nxqn3KKw9YnoKzcZvuArAfQ9GF3QG' },
+              { type: 'indirect',
+                hash: 'QmbJCNKXJqVK8CzbjpNFz2YekHwh3CSHpBA86uqYg3sJ8q' },
+              { type: 'indirect',
+                hash: 'QmVgSHAdMxFAuMP2JiMAYkB8pCWP1tcB9djqvq8GKAFiHi' }
+            ])
+          )
+      })
+
+      it('direct', function () {
+        return pin.ls({ type: 'direct' })
+          .then(out =>
+            expect(out).to.deep.eql([
+              { type: 'direct',
+                hash: 'QmbJCNKXJqVK8CzbjpNFz2YekHwh3CSHpBA86uqYg3sJ8q' }
+            ])
+          )
+      })
+
+      it('recursive', function () {
+        return pin.ls({ type: 'recursive' })
+          .then(out =>
+            expect(out).to.deep.eql([
+              { type: 'recursive',
+                hash: 'QmTAMavb995EHErSrKo7mB8dYkpaSJxu6ys1a6XJyB2sys' }
+            ])
+          )
+      })
+
+      it('indirect', function () {
+        return pin.ls({ type: 'indirect' })
+          .then(out =>
+            expect(out).to.deep.eql([
+              { type: 'indirect',
+                hash: 'QmTMbkDfvHwq3Aup6Nxqn3KKw9YnoKzcZvuArAfQ9GF3QG' },
+              { type: 'indirect',
+                hash: 'QmbJCNKXJqVK8CzbjpNFz2YekHwh3CSHpBA86uqYg3sJ8q' },
+              { type: 'indirect',
+                hash: 'QmVgSHAdMxFAuMP2JiMAYkB8pCWP1tcB9djqvq8GKAFiHi' }
+            ])
+          )
+      })
+    })
+  })
+
+  describe('rm', function () {
+    beforeEach(function () {
+      pin.clear()
+      return pin.add(pins.root)
+    })
+
+    it('a recursive pin', function () {
+      return pin.rm(pins.root)
+        .then(() => {
+          return Promise.all([
+            expectPinned(pins.root, false),
+            expectPinned(pins.mercuryWiki, false)
+          ])
+        })
+    })
+
+    it('a direct pin', function () {
+      pin.clear()
+      return pin.add(pins.mercuryDir, { recursive: false })
+        .then(() => pin.rm(pins.mercuryDir))
+        .then(() => expectPinned(pins.mercuryDir, false))
+    })
+
+    it('fails to remove an indirect pin', function () {
+      return pin.rm(pins.solarWiki)
+        .catch(err => expect(err).to.match(/is pinned indirectly under/))
+        .then(() => expectPinned(pins.solarWiki))
+    })
+
+    it('fails when an item is not pinned', function () {
+      return pin.rm(pins.root)
+        .then(() => pin.rm(pins.root))
+        .catch(err => expect(err).to.match(/is not pinned/))
+    })
+  })
+
+  describe('load', function () {
+    before(function () {
+      return pin.add(pins.root)
+    })
+
+    it('loads', function () {
+      pin.clear()
+      return pin.ls()
+        .then(ls => expect(ls.length).to.eql(0))
+        .then(() => pin.load())
+        .then(() => pin.ls())
+        .then(ls => expect(ls.length).to.be.gt(0))
```

Comment:
Is it not possible to assert on a known value in this test?  Same thing for the `it flushes` test.

Reply:
Yeah, I wasn't sure! My concern with hardcoded values is that it makes the tests fragile to changes in the default assets added during repo init, which these pin tests pick up.

Another approach would be to start each pin test with a clean slate by removing all pins. I did not go that route because of the added test overhead but looking back it's probably worth it. What do you think? Hardcode / naive `gt` checks / or cleaning the pins?

Reply:
I think a clean state is best in general, otherwise we could end up with tests that depend on the behaviour of other tests or the order in which they're run, etc.

I guess maybe try it and see how bad the overhead is?

Reply:
Turns out I was mistaken and in this case we can be explicit with the values because we are cleaning pins before/during tests. I was thinking of cli tests.